pages/shop.js

```diff
@@ -0,0 +1,165 @@
+import NavigationBar from '../frontend/components/NavigationBar';
+import React from 'react';
+import Router from 'next/router';
+import { Spinner, ToggleButtonGroup, ToggleButton } from 'react-bootstrap';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
+import { faArrowLeft, faExpand, faTimesCircle } from '@fortawesome/free-solid-svg-icons'
+import translate from '../frontend/components/translate.js';
+import { getItems } from '../frontend/actions/Items.js';
+import ShopCategories from '../frontend/components/ShopCategories.js';
+import CategoryItems from '../frontend/components/CategoryItems.js';
+import Search from '../frontend/components/Search.js';
+import SingleItemView from '../frontend/components/SingleItemView/SingleItemView';
+
+class ShoppingList extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      selectedValue: '1', isLoading: true, isUrgent: true, urgentOrUpcoming: [], categories: {}, searchItems: [], 
+      searchCategories: [], urgentItems: [], upcomingItems: [], currentItems: [], isCategoryList: true, isCategorySelected: false, isItemSelected: false, selectedCategory: null, 
+      backButton: false, shoppingNeeded: true, query: null
+    }
+    this.handleChange = this.handleChange.bind(this);
+  }
+  async componentDidMount() {
+    let urgentOrUpcoming = {};
+    let categories = {};
+    try {
+      let items = await getItems();
+      for (let item of items) {
+        if (item.reorder_level != null && item.reorder_level <= item.stock + 1) {
+          if (urgentOrUpcoming[item.category] == null) {
+            urgentOrUpcoming[item.category] = [];
+            categories[item.category] = 0;
+          }
+          urgentOrUpcoming[item.category].push(item);
+          categories[item.category]++;
+        }
+      }
+      if (categories.length == 0) {
+        this.setState({shoppingNeeded: false});
+      } else {
+        this.setState({ isLoading: false, urgentOrUpcoming: urgentOrUpcoming, categories: categories });
+      }
+    } catch (e) {
+      console.error(e);
+    }
+  }
+  handleChange(value) {
+    this.setState({ selectedValue: value });
+    let foundMatches = [];
+    switch (value) {
+      case 1:
+        this.setState({ isUrgent: true, currentItems: this.state.urgentItems }); 
+        break;
+      case 2:
+        this.setState({ isUrgent: false, currentItems: this.state.upcomingItems });
+        break; 
+    }
+  }
+  selectCategory = (category) => {
+    let urgentOrUpcoming = this.state.urgentOrUpcoming[category];
+    let urgentItems = [];
+    let upcomingItems = [];
+    for (let item of urgentOrUpcoming) {
+      if (item.reorder_level <= item.stock) {
+        urgentItems.push(item);
+      } else {
+        upcomingItems.push(item);
+      }
```

Comment:
We may need to change this algorithm in the future, as I think some items are neither upcoming nor urgent. We'll figure out the specifics later though.

Reply:
Yeah sorry its a little confusing. If you look at line 30-36, it shows how urgentOrUpcoming is created. 
```
     if (item.reorder_level != null && item.reorder_level <= item.stock + 1) {
       if (urgentOrUpcoming[item.category] == null) {
          urgentOrUpcoming[item.category] = [];
          categories[item.category] = 0;
       }
       urgentOrUpcoming[item.category].push(item);
       categories[item.category]++;
     }
```
I did this so that <ShopCategories> would only receive the items and categories that are either urgent or upcoming (as opposed to everything). Once the category gets selected though, I sort urgentOrUpcoming into an urgent array and an upcoming array. 
```
    for (let item of urgentOrUpcoming) {
      if (item.reorder_level <= item.stock) {
        urgentItems.push(item);
      } else {
        upcomingItems.push(item);
      }
    } 
```
So I could have had an else if (reorder_level == item.stock + 1) but everything else in urgentOrUpcoming is going to be in upcoming no matter what. Also for reference the Notion says that  urgent <= stock, and stock < upcoming < stock + 2.

Reply:
Ok that makes sense. I think in the call we discussed we may want to change upcoming to be within a certain percentage from reorder_level rather than within 2 of the reorder_level, but we never came to a conclusion. So its good you implemented it in the same way it worked as the original app / whats on Notion. It should be easy to change in the future. Nice job!